src/icu_win.cc

```diff
@@ -0,0 +1,238 @@
+    g_unavailable.store(true, std::memory_order_relaxed);
+    return Unavailable();
+  }
+
+  // Store the function pointers
+  g_ucal_getHostTimeZone.store(ucal_getHostTimeZoneRef,
+                               std::memory_order_relaxed);
+  g_ucal_getTimeZoneIDForWindowsID.store(ucal_getTimeZoneIDForWindowsIDRef,
+                                         std::memory_order_relaxed);
+
+  return {true, ucal_getHostTimeZoneRef, ucal_getTimeZoneIDForWindowsIDRef};
+}
+
+// Convert wchar_t array (UTF-16) to UTF-8 string
+std::string Utf16ToUtf8(const wchar_t* ptr, size_t size) {
+  if (size > std::numeric_limits<int>::max()) {
+    return std::string();
+  }
+  const int chars_len = static_cast<int>(size);
+  int num_bytes_in_utf8 = 0;
+  {
+    std::array<char, 32> buffer;
+    num_bytes_in_utf8 =
+        ::WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, ptr, chars_len,
+                              buffer.data(), buffer.size(), nullptr, nullptr);
+    if (num_bytes_in_utf8 <= buffer.size()) {
+      return std::string(buffer.data(), num_bytes_in_utf8);
+    }
+  }
+  auto buffer = std::unique_ptr<char[]>(new char[num_bytes_in_utf8]);
+  const int num_written_bytes =
+      ::WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, ptr, chars_len,
+                            buffer.get(), num_bytes_in_utf8, nullptr, nullptr);
+  if (num_written_bytes != num_bytes_in_utf8) {
+    return std::string();
+  }
+  return std::string(buffer.get(), num_written_bytes);
+}
+
+}  // namespace
+
+std::string GetWindowsLocalTimeZone() {
+  const auto icu = GetIcuFunctions();
+  if (!icu.available) {
+    return std::string();
+  }
+
+  std::int32_t length = 0;
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  // Try ucal_getHostTimeZone first (available on Windows 11+)
+  if (icu.ucal_getHostTimeZone != nullptr) {
+    std::array<UChar, 32> buffer;
+    length = icu.ucal_getHostTimeZone(buffer.data(), buffer.size(), &status);
+    if (U_SUCCESS(status) && length >= 0) {
+      return Utf16ToUtf8(buffer.data(), length);
+    }
+    if (status == U_BUFFER_OVERFLOW_ERROR && length > 0) {
+      const int buffer_size = length + 1;  // +1 for null terminator
+      auto buffer = std::unique_ptr<UChar[]>(new UChar[buffer_size]);
+      status = U_ZERO_ERROR;
+      length = icu.ucal_getHostTimeZone(buffer.get(), buffer_size, &status);
+      if (U_SUCCESS(status) && length >= 0) {
+        return Utf16ToUtf8(buffer.get(), length);
+      }
+    }
+  }
+
+  DYNAMIC_TIME_ZONE_INFORMATION info = {};
+  if (::GetDynamicTimeZoneInformation(&info) == TIME_ZONE_ID_INVALID) {
+    return std::string();
+  }
+  {
+    std::array<UChar, 32> buffer;
+    status = U_ZERO_ERROR;
+    length = icu.ucal_getTimeZoneIDForWindowsID(info.TimeZoneKeyName, -1,
+                                                nullptr, buffer.data(),
+                                                buffer.size(), &status);
+    if (U_SUCCESS(status) && length >= 0) {
```

Comment:
Does `ucal_getTimeZoneIDForWindowsID()` ever return a negative value when it succeeds?

Reply:
Fair enough. It's whether we trust lower level APIs (ucal_getTimeZoneIDForWindowsID() in this case) or not, but let's remove this redundant check.